Finished applying fixes for jbmc for assignment consistency

thk123 opened the pull request:
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message). TODO
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

kroening committed:
symex: check type consistency of assignments

kroening committed:
expr_initializert: fix type of enum constants

smowton committed:
Ensure member_exprt's type matches the addressed component

Previously we created the expression with the partial type given at the
access site (i.e. as part of the
get/setfield bytecode), but this lacked any generic type information,
which the actual component *did* have.
The member expression could therefore be type-inconsistent with the
object is addressed. Since it's no longer
necessary to make the expression approximately and later fix it up in
typecheck_expr, let's just get it right
the first time, including giving it the right type.

smowton committed:
Java method converter: don't hold a symbol table reference across table add operations

This is generally illegal (but is tricky to trigger, hence no test)

smowton committed:
Java method converter: ensure type consistency of generated function calls

This both ensures that the reference to the callee precisely matches the callee's symbol type,
and that parameters match. For example, previously the callee could specifically take a List<Integer>,
but the type given at the callsite might just be `callee(List)`, omitting the generic type information,
leading to a pointer changing type without a cast.

smowton committed:
Java pointer casts: ensure type consistency, even when casting to a base type

This is necessary because a generic qualifier is recorded on the pointer, while the direct member has an
unqualified type.

smowton committed:
Remove virtual functions: cast argument types as necessary

For example, when List<E> extends the Collection<E> interface, its overrides of methods accepting an element,
such as add(E), are given in terms of its own E, not that of Collection. Therefore we must cast from Collection::E*
to List::E* when using that particular override.

Only typecast if the parameter types are distinct

smowton committed:
Allocate objects: cast when pointer types differ

It already tried to do this, but doesn't take account for pointers that encode generic information:
in such cases the underlying object type can be equal but the pointers differ, and a cast is necessary.

Commit:
Reduce type equality check within symex

Once all the cases are fixed, this can be returned to a regular equality

Commit:
Increased description in test to make it clear what it is testing

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -114,16 +114,36 @@ static void create_static_function_call(
+        "those arguments must be pointer-typed");
+    }
+
+    call_args[i].make_typecast(need_type);
```
This seems pointless - given the above `type_eq` the difference could now only be tag/symbol vs non-tag/symbol types.

smowton replied:
Not so, that only checked that the types are either `type_eq` or are pointers

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1531,7 +1550,7 @@ code_blockt java_bytecode_convert_methodt::convert_instructions(
     {
       PRECONDITION(op.size() == 1 && results.size() == 1);
-      results[0]=java_bytecode_promotion(to_member(op[0], arg0));
+      results[0]=java_bytecode_promotion(to_member(op[0], arg0, ns));
```
This PR will need per-commit clang-format cleanup.

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -140,9 +140,8 @@ static void create_static_function_call(
         call_args[i].type().id() == ID_pointer,
         "where overriding function argument types differ, "
         "those arguments must be pointer-typed");
+      call_args[i].make_typecast(need_type);
```
Ah, please squash this into the commit where I questioned its utility.

tautschnig commented:
Oops, missed the do-not-review label.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
 }
 
-static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
+static member_exprt to_member(
```
documentation here would help

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
 
-static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
+static member_exprt to_member(
+  const exprt &pointer, const exprt &fieldref, const namespacet &ns)
```
:pick: fieldref is not a good name

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
+    pointer, java_reference_type(class_type));
 
-  dereference_exprt obj_deref=checked_dereference(pointer2, class_type);
+  const irep_idt &component_name = fieldref.get(ID_component_name);
```
Is fieldref assumed to be a `member_exprt`? It would be better to make that explicit.

smowton replied:
It isn't `member_exprt`, but rather the nameless expression that describes a `getfield` or `setfield` operand produced by the Java bytecode parser

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
+      to_struct_type(ns.follow(accessed_object.type()));
+    auto component = object_type.get_component(component_name);
+
+    if(component.is_not_nil())
```
:pick: I would prefer if the loop was expressed as `while(component.is_nil())` rather than `while(1)`

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
+      "infer_opaque_type_fields should guarantee that a member access has a "
+      "corresponding field");
+
+    accessed_object = member_exprt(accessed_object, components.front());
```
:bulb: ideally there should be a function giving the parent class from a java class type, that would make this piece of code clearer.

tautschnig replied:
Dumb question: what is it that assures us that we always need to look at the first component? I get that because of Java not permitting multiple inheritance if anything, it must be the first one. Is it because everything inherits at least from `java.lang.object` that we know that this first component must be a struct type?

smowton replied:
Yep, invariant of Java object layout: every class has at least one member, and the first one is either `@class_identifier` (java.lang.Object, or any stub class whose position in the class hierarchy is unknown) or its single superclass.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
+
+    // Component not present: check the immediate parent type.
+
+    const auto &components = object_type.components();
```
:pick: Looks like `resolve_inherited_componentt` can be adapted to do this search

thk123 replied:
That said - this is just moving code from the typecheck so will clean up in a follow up PR/commit

Commit:
Fixing unit tests wrt new GOTO format

Specifically - be impartial to deref/cast/address of combinations on the
lhs.

Be impartial to direct assignments without casts
Be impartial to casts before assigning to address of another variable

thk123 commented:
So I couldn't reproduce the failure locally, seemingly because `ctest` does not run the `symex-driven-lazy-loading-expected-failure` target correctly (finds 0 tests to run). I fixed this by removing `-s;symex-driven-loading` from the `CMakeLists.txt` in the relevant regression folder and now get the tests. 

I don't understand this change, so if someone does and wants to PR it, that would have my approval. Otherwise, I will PR it after I've got this working.

tautschnig commented:
@thk123 Are you working off of a branch that has 3313f25a9fb5f667292d49bb57a359253ee19ffc?

thk123 commented:
@tautschnig thanks for the tip @smowton just pointed me towards that - now got the failure locally. It is to do with an invalid pop and only with symex driven lazy loading so will keep digging.

thk123 commented:
Tracked down the problem with the malformed pop (remove returns had already run on a function, so swapping out the type for the type in the symbol table meant no returns). Just checking locally that passes all the tests, then will push up here to verify on CI and then will create a TG bump since plausible it will break things there too.

Commit:
Resolve issue with return type being erased

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 8366f3b5e47c11a4a1f2790057ba4b23f95ecae8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97821114
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

thk123 commented:
Even when the compile errors are fixed, this fails in TG for non-equal types in symex - this looks to be to do with the custom stubbing - manual bump incoming.

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -647,13 +647,13 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
 
-  const exprt typed_pointer = typecast_exprt::conditional_cast(
-    pointer, java_reference_type(class_type));
+  const exprt typed_pointer =
```
This commit contains format fixups that ought to be squashed into a previous commit

smowton reviewed (approved):
Can't approve my own code of course, but I agree with @thk123 and @kroening's parts

thk123 commented:
TG bump looks to be passing - will address existing comments and tidy up, then should be ready for a final review from @tautschnig

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 050dbec33c530bcefd7325ea55d727c301951901).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98066474

tautschnig commented:
@thk123 Just send a ping when you're done with cleanup work.

tautschnig commented:
@thk123 Actually there isn't anything new that I hadn't looked at in the meantime, so just do the cleanup, merge, and update TG accordingly. Thanks!

thk123 commented:
Am now happy with this, so review away @tautschnig 
I would particularly like some feedback on the [`root_object`](https://github.com/diffblue/cbmc/pull/3822/files#diff-ac8338bc55f009ec0eefcd32d58578b5R26) method in the unit tests - it feels like this logic may (must?) exist elsewhere already.

tautschnig commented:
@thk123 re `root_object`: there is `object_descriptor_exprt`, which can be passed an `exprt` via its `build` method and then you get a `.root_object()` - but I need to re-read your code to see whether this does exactly what you need.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -642,20 +642,39 @@ static irep_idt get_if_cmp_operator(const irep_idt &stmt)
+  // may in fact belong to any ancestor type.
+  while(1)
+  {
+    struct_typet object_type =
```
I'd use `const struct_typet &object_type = ...` here.

thk123 replied:
This code is in fact simply moved - I'd prefer to follow this PR up with a clean up PR as this whole block I think can be replaced by `resolve_inherited_componentt`

tautschnig replied:
Sounds good!

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -114,16 +114,35 @@ static void create_static_function_call(
+        call_args[i].type().id() == ID_pointer,
+        "where overriding function argument types differ, "
+        "those arguments must be pointer-typed");
+      call_args[i].make_typecast(need_type);
```
Avoid using `make_typecast` (it's now deprecated), do `call_args[i] = typecast_exprt(call_args[i], need_type);` instead.

tautschnig commented on jbmc/unit/java-testing-utils/require_goto_statements.cpp:
```diff
@@ -11,11 +11,47 @@ Author: Diffblue Ltd.
-#include <util/expr_iterator.h>
 #include <goto-programs/goto_functions.h>
+#include <goto-programs/show_goto_functions.h>
+#include <iostream>
```
Is that a debugging left-over?

tautschnig commented on jbmc/unit/java-testing-utils/require_goto_statements.cpp:
```diff
@@ -11,11 +11,47 @@ Author: Diffblue Ltd.
+/// by skipping over type casts and removing balanced dereference/address_of
+/// operations
+optionalt<symbol_exprt>
+root_object(const exprt &lhs_expr, const symbol_tablet &symbol_table)
```
As said in the comment, there is `object_descriptor_exprt`. I think it may be necessary to extend `object_descriptor_exprt::build_object_descriptor_rec` with code handling paired `dereference_exprt/address_of_exprt`, but then it should do the trick for you. I believe all the code that this needs in addition is (some variant of):
```
else if(expr.id() == ID_dereference)
{
  const dereference_exprt &deref = to_dereference_expr(expr);
  if(deref.pointer().id() == ID_address_of)
  {
    const address_of_exprt &address_of = to_address_of_expr(deref.pointer());
    dest.object() = address_of.object();
    build_object_descriptor_rec(ns, address_of.object(), dest);
  }
}
```

thk123 replied:
Yes I saw the `object_descriptor_exprt` and saw that it didn't handle the ref/deref. In the interest of not wanting to break anything else that uses this, I will attempt this refactor in a follow up PR too if that is acceptable to you.

tautschnig replied:
Sure!

tautschnig replied:
I'll post a PR in a few minutes.

tautschnig replied:
See #3898.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -41,11 +41,12 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <algorithm>
 #include <functional>
-#include <unordered_set>
+#include <goto-programs/remove_returns.h>
```
I'd prefer not mixing our headers with STL includes.

thk123 commented:
> what is it that assures us that we always need to look at the first component?

It probably isn't a great assumption - it is _true_ because in `java_bytecode_convert_class` the base class component is always added first (line 370) and then additional fields (line 458). As with other comment - would prefer to address in follow up PR using `resolve_inherited_componentt`

thk123 commented:
Seem to have done enough rebases to make clang-format go away...

thk123 commented:
@tautschnig addressed minor comments, larger changes I would like to address in a follow up PR. If that's OK with you, this looks good to merge once CI passes.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5a9bf1ad65350f5a0fe074bf84f730aaab5ba88d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98303689
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f0c520e8aec055afddbebc95457e70bdc7eb521a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98304937
